File: ChatDemo-UI3.0/ChatDemo-UI3.0/Code/Controllers/Chats/Views/ChatCell/EMChatVideoBubbleView.m
/************************************************************
 *  * Hyphenate
 * __________________
 * Copyright (C) 2016 Hyphenate Inc. All rights reserved.
 *
 * NOTICE: All information contained herein is, and remains
 * the property of Hyphenate Inc.
 */

#import "EMChatVideoBubbleView.h"

#import "EMMessageModel.h"

#define MAX_SIZE 250

@interface EMChatVideoBubbleView ()

@property (strong, nonatomic) UIButton *videoPlayButton;

@end

@implementation EMChatVideoBubbleView

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        self.layer.masksToBounds = YES;
        self.contentMode = UIViewContentModeScaleAspectFill;
        self.layer.cornerRadius = 10.f;
        
        self.videoPlayButton = [UIButton buttonWithType:UIButtonTypeCustom];
        UIImage *backgroundImage = [UIImage imageNamed:@"Icon_Play"];
        [self.videoPlayButton setBackgroundImage:backgroundImage forState:UIControlStateNormal];
        [self addSubview:self.videoPlayButton];
    }
    return self;
}

- (void)layoutSubviews
{
    [super layoutSubviews];
    CGFloat width = 50.0f;
    CGFloat height = 50.0f;
    CGFloat x = self.frame.size.width/2 - width/2;
    CGFloat y = self.frame.size.height/2 - height/2;
    [self.videoPlayButton setFrame:CGRectMake(x, y, width, height)];
}

- (CGSize)sizeThatFits:(CGSize)size
{
    CGSize retSize;
    EMVideoMessageBody *body = (EMVideoMessageBody*)self.model.message.body;
    if (self.model.message.ext) {
        retSize = CGSizeMake(0, 0);
    } else {
        retSize = body.thumbnailSize;
    }
    if (retSize.width == 0 || retSize.height == 0) {
        retSize.width = MAX_SIZE;
        retSize.height = MAX_SIZE;
    }
    if (retSize.width > retSize.height) {
        CGFloat height =  MAX_SIZE / retSize.width  *  retSize.height;
        retSize.height = height;
        retSize.width = MAX_SIZE;
    }else {
        CGFloat width = MAX_SIZE / retSize.height * retSize.width;
        retSize.width = width;
        retSize.height = MAX_SIZE;
    }
    if (self.model.message.ext) {
        retSize.height = MAX_SIZE / 4 * 3;
    }
    return retSize;
}

- (void)setModel:(EMMessageModel*)model;
{
    [super setModel:model];
    
    EMVideoMessageBody *videoBody = (EMVideoMessageBody *)model.message.body;
    if ([videoBody.thumbnailLocalPath length] > 0) {
        NSData *thumbnailImageData = [NSData dataWithContentsOfFile:videoBody.thumbnailLocalPath];
        if (thumbnailImageData.length) {
            self.backImageView.image = [UIImage imageWithData:thumbnailImageData];
        }
    } else {
        [self.backImageView sd_setImageWithURL:[NSURL URLWithString:videoBody.thumbnailRemotePath] placeholderImage:nil];
    }
}

+ (CGFloat)heightForBubbleWithMessageModel:(EMMessageModel *)model
{
    CGSize retSize;
    EMVideoMessageBody *body = (EMVideoMessageBody*)model.message.body;
    if (model.message.ext) {
        retSize = CGSizeMake(0, 0);
    } else {
        retSize = body.thumbnailSize;
    }
    if (retSize.width == 0 || retSize.height == 0) {
        retSize.width = MAX_SIZE;
        retSize.height = MAX_SIZE;
    }
    if (retSize.width > retSize.height) {
        CGFloat height =  MAX_SIZE / retSize.width  *  retSize.height;
        retSize.height = height;
        retSize.width = MAX_SIZE;
    }else {
        CGFloat width = MAX_SIZE / retSize.height * retSize.width;
        retSize.width = width;
        retSize.height = MAX_SIZE;
    }
    if (model.message.ext) {
        retSize.height = MAX_SIZE / 4 * 3;
    }
    
    return retSize.height;
}

@end

File: swift-demo/Hyphenate-Demo-Swift/Code/Controllers/Chats/Views/ChatCell/EMChatVideoBubbleView.swift
//
//  EMChatVideoBubbleView.swift
//  Hyphenate-Demo-Swift
//
//  Created by dujiepeng on 2017/6/23.
//  Copyright © 2017 dujiepeng. All rights reserved.
//

import UIKit
import Hyphenate
import SDWebImage

class EMChatVideoBubbleView: EMChatBaseBubbleView {
    var videoButton: UIButton!
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        layer.masksToBounds = true
        contentMode = UIViewContentMode.scaleAspectFill
        layer.cornerRadius = 10
        
        videoButton = UIButton(type: UIButtonType.custom)
        videoButton.setBackgroundImage(UIImage(named:"Icon_Play"), for: UIControlState.normal)
        videoButton.isUserInteractionEnabled = false
        addSubview(videoButton)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        videoButton.frame = CGRect(x: (width() - 50) / 2, y: (height() - 50) / 2, width: 50, height: 50)
    }
    
    override func sizeThatFits(_ size: CGSize) -> CGSize {
        var retSize: CGSize = CGSize.zero
        let body = _model?.message?.body as! EMVideoMessageBody
        if _model?.message?.ext != nil {
            retSize = CGSize(width: 0, height: MAX_SIZE / 4 * 3)
        }else {
            retSize = body.thumbnailSize
        }
        
        if retSize.width == 0 || retSize.height == 0 {
            retSize = CGSize(width: MAX_SIZE, height: MAX_SIZE)
        }
        
        if retSize.width > retSize.height {
            let height = MAX_SIZE / retSize.width * retSize.height
            retSize.height = height
            retSize.width = MAX_SIZE
        } else {
            let width = MAX_SIZE / retSize.height * retSize.width
            retSize.height = MAX_SIZE
            retSize.width = width
        }
        
        return retSize
    }
    
    override func set(model: EMMessageModel) {
        super.set(model: model)
        let body = _model?.message?.body as! EMVideoMessageBody
        if body.thumbnailLocalPath != nil {
            let thumnailImageData = NSData.init(contentsOfFile: body.thumbnailLocalPath)
            if thumnailImageData != nil {
                _backImageView?.image = UIImage.init(data: thumnailImageData! as Data)
            }
        }else {
            _backImageView?.sd_setImage(with: URL.init(string: body.remotePath), placeholderImage: nil)
        }
    }
    
    class func heightForBubble(withMessageModel model: EMMessageModel) -> CGFloat {
        var retSize: CGSize = CGSize.zero
        let body = model.message?.body as! EMVideoMessageBody
        if model.message?.ext != nil {
            retSize = CGSize(width: 0, height: MAX_SIZE / 4 * 3)
        }else {
            retSize = body.thumbnailSize
        }
        
        if retSize.width == 0 || retSize.height == 0 {
            retSize = CGSize(width: MAX_SIZE, height: MAX_SIZE)
        }
        
        if retSize.width > retSize.height {
            let height = MAX_SIZE / retSize.width * retSize.height
            retSize.height = height
            retSize.width = MAX_SIZE
        } else {
            let width = MAX_SIZE / retSize.height * retSize.width
            retSize.height = MAX_SIZE
            retSize.width = width
        }
        
        return retSize.height
    }
}
